File: lesibuy-client/src/app/services/order.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AuthService } from './auth.service';

export interface OrderItem {
  id?: number;
  productId?: number;
  productName: string;
  productDescription?: string;
  productImageUrl?: string;
  productCondition?: string;
  sellerType?: string;
  specifications?: string;
  quantity: number;
  unitPrice: number;
  subTotal: number;
}

export interface Order {
  id: number;
  createdAt: string;
  fullName: string;
  email: string;
  phone: string;

  shippingAddress: string;
  shippingCity: string;
  shippingPostalCode: string;

  billingSameAsShipping: boolean;
  billingAddress?: string;
  billingCity?: string;
  billingPostalCode?: string;

  paymentMethod: string;
  totalAmount: number;
  status: string;
  items: OrderItem[];
}

export interface CreateOrderItemRequest {
  productId: number;
  quantity: number;
}

export interface CreateOrderRequest {
  fullName: string;
  email?: string;
  phone?: string;

  shippingAddress: string;
  shippingCity: string;
  shippingPostalCode: string;

  billingSameAsShipping: boolean;
  billingAddress?: string;
  billingCity?: string;
  billingPostalCode?: string;

  paymentMethod?: string;
  items: CreateOrderItemRequest[];
}

export interface CreateOrderResponse {
  id: number;
  createdAt: string;
  fullName: string;
  email: string;
  phone: string;

  shippingAddress: string;
  shippingCity: string;
  shippingPostalCode: string;

  billingSameAsShipping: boolean;
  billingAddress?: string;
  billingCity?: string;
  billingPostalCode?: string;

  paymentMethod: string;
  totalAmount: number;
  status: string;
  items: OrderItem[];
}

@Injectable({
  providedIn: 'root'
})
export class orderservice {
  private apiUrl = '/api/orders';

  constructor(
    private http: HttpClient,
    private authService: AuthService
  ) { }

  getorders(): Observable<Order[]> {
    return this.http.get<Order[]>(this.withAccessToken(this.apiUrl));
  }

  getMyorders(): Observable<Order[]> {
    return this.http.get<Order[]>(this.withAccessToken(this.apiUrl));
  }

  getOrderById(id: number): Observable<Order> {
    return this.http.get<Order>(this.withAccessToken(`${this.apiUrl}/${id}`));
  }

  createOrder(payload: CreateOrderRequest): Observable<CreateOrderResponse> {
    return this.http.post<CreateOrderResponse>(this.withAccessToken(this.apiUrl), payload);
  }

  cancelOrder(id: number): Observable<{ message: string }> {
    return this.http.put<{ message: string }>(
      this.withAccessToken(`${this.apiUrl}/${id}/cancel`),
      {}
    );
  }

  private withAccessToken(url: string): string {
    const token = this.authService.getToken();

    if (!token) {
      return url;
    }

    const separator = url.includes('?') ? '&' : '?';
    return `${url}${separator}access_token=${encodeURIComponent(token)}`;
  }
}

File: LesiBuy/LesiBuy.Application/Services/OrderService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using LesiBuy.Application.Dtos;
using LesiBuy.Domain.Entities;
using LesiBuy.Domain.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace LesiBuy.Application.Services
{
    public class OrderService : IOrderService
    {
        private readonly IUnitOfWork _uow;

        public OrderService(IUnitOfWork uow)
        {
            _uow = uow;
        }

        public async Task<OrderDto> CreateOrderAsync(CreateOrderDto dto, int? userId = null)
        {
            var billingAddress = dto.BillingSameAsShipping ? dto.ShippingAddress : dto.BillingAddress;
            var billingCity = dto.BillingSameAsShipping ? dto.ShippingCity : dto.BillingCity;
            var billingPostalCode = dto.BillingSameAsShipping ? dto.ShippingPostalCode : dto.BillingPostalCode;

            var order = new Order
            {
                FullName = dto.FullName,
                Email = dto.Email,
                Phone = dto.Phone,

                ShippingAddress = dto.ShippingAddress,
                ShippingCity = dto.ShippingCity,
                ShippingPostalCode = dto.ShippingPostalCode,

                BillingSameAsShipping = dto.BillingSameAsShipping,
                BillingAddress = billingAddress,
                BillingCity = billingCity,
                BillingPostalCode = billingPostalCode,

                PaymentMethod = dto.PaymentMethod,
                CreatedAt = DateTime.UtcNow,
                TotalAmount = 0,
                UserId = userId ?? 0,
                Status = "Pending",
                Items = new List<OrderItem>()
            };

            decimal total = 0;

            foreach (var item in dto.Items)
            {
                var product = await _uow.Products.GetByIdAsync(item.ProductId);

                if (product == null)
                    throw new Exception($"Product with ID {item.ProductId} not found.");

                var subTotal = product.Price * item.Quantity;

                order.Items.Add(new OrderItem
                {
                    ProductId = product.Id,
                    ProductName = product.Name,
                    ProductDescription = product.Description,
                    ProductImageUrl = product.ImageUrl,
                    ProductCondition = product.Condition,
                    SellerType = product.SellerType,
                    Specifications = product.Specifications,
                    UnitPrice = product.Price,
                    Quantity = item.Quantity,
                    SubTotal = subTotal
                });

                total += subTotal;
            }

            order.TotalAmount = total;

            await _uow.Repository<Order>().AddAsync(order);
            await _uow.CompleteAsync();

            return MapOrderToDto(order);
        }

        public async Task<IReadOnlyList<OrderDto>> GetAllOrdersAsync()
        {
            var orders = await _uow.Repository<Order>()
                .Query()
                .Include(o => o.Items)
                .OrderByDescending(o => o.CreatedAt)
                .ToListAsync();

            return orders.Select(MapOrderToDto).ToList();
        }

        public async Task<IReadOnlyList<OrderDto>> GetOrdersByUserIdAsync(int userId)
        {
            var orders = await _uow.Repository<Order>()
                .Query()
                .Where(o => o.UserId == userId)
                .Include(o => o.Items)
                .OrderByDescending(o => o.CreatedAt)
                .ToListAsync();

            return orders.Select(MapOrderToDto).ToList();
        }

        public async Task<OrderDto?> GetOrderByIdAsync(int id)
        {
            var order = await _uow.Repository<Order>()
                .Query()
                .Include(o => o.Items)
                .FirstOrDefaultAsync(o => o.Id == id);

            if (order == null)
                return null;

            return MapOrderToDto(order);
        }

        public async Task<bool> CancelOrderAsync(int orderId, int userId)
        {
            var order = await _uow.Repository<Order>()
                .Query()
                .FirstOrDefaultAsync(o => o.Id == orderId && o.UserId == userId);

            if (order == null)
                return false;

            if (!string.Equals(order.Status, "Pending", StringComparison.OrdinalIgnoreCase))
                return false;

            order.Status = "Cancelled";
            await _uow.CompleteAsync();

            return true;
        }

        private static OrderDto MapOrderToDto(Order order)
        {
            return new OrderDto
            {
                Id = order.Id,
                FullName = order.FullName,
                Email = order.Email,
                Phone = order.Phone,

                ShippingAddress = order.ShippingAddress,
                ShippingCity = order.ShippingCity,
                ShippingPostalCode = order.ShippingPostalCode,

                BillingSameAsShipping = order.BillingSameAsShipping,
                BillingAddress = order.BillingAddress,
                BillingCity = order.BillingCity,
                BillingPostalCode = order.BillingPostalCode,

                PaymentMethod = order.PaymentMethod,
                TotalAmount = order.TotalAmount,
                CreatedAt = order.CreatedAt,
                Status = order.Status,

                Items = order.Items.Select(i => new OrderItemDto
                {
                    ProductId = i.ProductId,
                    ProductName = i.ProductName,
                    ProductDescription = i.ProductDescription,
                    ProductImageUrl = i.ProductImageUrl,
                    ProductCondition = i.ProductCondition,
                    SellerType = i.SellerType,
                    Specifications = i.Specifications,
                    UnitPrice = i.UnitPrice,
                    Quantity = i.Quantity,
                    SubTotal = i.SubTotal
                }).ToList()
            };
        }
    }
}